File: LongestSubstringWithoutRepeatingCharacters.c
int lengthOfLongestSubstring(char * s){
    int maxLen = 0;
    int start = 0;
    int map[256] = {0}; // ASCII character map
    for(int end = 0; s[end] != '\0'; end++) {
        char ch = s[end];
        while(map[(unsigned char)ch]) {
            map[(unsigned char)s[start]] = 0;
            start++;
        }
        map[(unsigned char)ch] = 1;
        int currLen = end - start + 1;
        if(currLen > maxLen) maxLen = currLen;
    }
    return maxLen;
}

File: LongestSubstringWithoutRepeatingCharacters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int left = 0, right = 0, maxLen = 0;
        while (right < s.length()) {
            while (set.contains(s.charAt(right))) {
                set.remove(s.charAt(left));
                left++;
            }
            set.add(s.charAt(right));
            maxLen = Math.max(maxLen, right - left + 1);
            right++;
        }
        return maxLen;
    }
}